### ISSUE TITLE
[clang-tidy] `bugprone-unchecked-optional-access` warns after `ASSERT_TRUE(optional.has_value())` in GoogleTest

### ISSUE THREAD

**Issue by Stee-T** [llvm/llvm-project#181737]
**Description**

`bugprone-unchecked-optional-access` emits a warning when accessing an `std::optional` value inside a GoogleTest unit test, even if the optional is checked using a fatal GoogleTest assertion that terminates the test on failure.

Because `ASSERT_*` macros abort the current test function when the condition is false, execution cannot reach the `.value()` call unless the optional contains a value. Therefore this access should be considered guarded.

**Minimal example**

```cpp
#include <optional>
#include <gtest/gtest.h>

TEST( Example, OptionalChecked ) {
  std::optional< int > MyOptional = 42069;

  ASSERT_TRUE( MyOptional.has_value() );  // Fatal: aborts test if false
  EXPECT_EQ( MyOptional.value(), 42069 ); // clang-tidy warns here
}
```

**Actual behavior**

clang-tidy reports an "unchecked access to optional value" warning on the `.value()` call.

**Expected behavior**

No warning should be emitted after a fatal GoogleTest assertion that checks `has_value()`, since control flow cannot continue when the condition is false.

**Suggested improvement**

Treat GoogleTest fatal assertions (e.g. `ASSERT_TRUE(optional.has_value())`) as control-flow-terminating guards, similar to how other terminating constructs are modeled.

**Comment by jvoung** [comment]
@BaLiKfromUA would definitely appreciate help here =)

Previously, we tried to add an option to filter out gtest source files https://github.com/llvm/llvm-project/pull/115051 but was stuck, but yes, now there is also an example in abseil/unchecked-statusor-access transfer functions for built-in support. I think `ASSERT_TRUE` is the most common one. There are other forms that are equivalent like `ASSERT_NE(..., nullopt)` or `ASSERT_THAT(...)` that are less common, but are still sometimes used.

Do you have a pointer to the "However, the gtest maintainers have declined, stating that clang-tidy should adapt instead."?

**Comment by BaLiKfromUA** [comment]
> Do you have a pointer to the "However, the gtest maintainers have declined, stating that clang-tidy should adapt instead."?

I've made a very cursory look but this is an impression that I got based on https://github.com/google/googletest/pull/3676#issuecomment-1030228060 , https://github.com/google/googletest/pull/3983 , and https://github.com/google/googletest/pull/4091

Again, it was not about `bugprone-unchecked-optional-access` and I do apologies if I misinterpreted responses from maintainers.

Anyway, thanks for giving a green light! I will take a look

**Comment by BaLiKfromUA** [comment]
@jvoung I created the initial patch to support `ASSERT_TRUE`. 

> There are other forms that are equivalent like ASSERT_NE(..., nullopt) or ASSERT_THAT(...) that are less common, but are still sometimes used.

Next I am looking on `ASSERT_THAT` support but probably I would prefer to send it as a separate PR.

Will continue working on it after initial code review of current PR since I have some doubts about the code.
